qualtran/_infra/composite_bloq.py

```diff
@@ -66,6 +66,11 @@
 canonicalize and return `SoquetT`.
 """
 
+_ConnectionType = TypeVar('_ConnectionType', bound=np.generic)
+
+ConnectionT = Union[Connection, NDArray[_ConnectionType]]
```

Comment:
NDArray is typed so that anything in an NDArray has to be a np.generic.  So, if you have NDArray[X], then X has to be a np.generic.  Otherwise, mypy generates a type error.  This issue happens when you jam something that's not a numpy type into a numpy array.

I don't know anything about _ConnectionType.